### PR TITLE
Issue #1191 by JrgenGNielsen: Clear page cache if changed maintainanc…

### PR DESCRIPTION
**Pull request by JrgenGNielsen**
…e status

### CODE REVIEW

File: ding_frontend.module [hunk 1]
```diff
@@ -112,3 +112,24 @@ function ding_frontend_query_alter($query) {
...
   }
 }
+
+/**
+ * Implements hook_boot().
+ *
+ * Clear page cache if maintainance status has changed.
+ */
+function ding_frontend_boot() {
+  $ding_maintenance_toggle = variable_get('ding_maintenance_toggle', 0);
```
Comment (kasperg) [inline ding_frontend.module]: Variables should be prefixed with the name of the responsible module e.g. `ding_frontend_maintenance_toggle`.

File: ding_frontend.module [hunk 2]
```diff
@@ -112,3 +112,24 @@ function ding_frontend_query_alter($query) {
...
+    if (!$ding_maintenance_toggle) {
+      cache_clear_all(NULL, 'cache_page');
+      variable_set('ding_maintenance_toggle', 1);
+    }
+  }
+  else {
+    if ($ding_maintenance_toggle) {
+      cache_clear_all(NULL, 'cache_page');
+      variable_set('ding_maintenance_toggle', 0);
+    }
```
Comment (kasperg) [inline ding_frontend.module]: If you revisit this code then consider reducing nesting.

**Comment by kasperg** [comment]
@JrgenGNielsen I have reviewed this and I agree that it solves the problem. I am unsure about the approach taken.

- This code will be run for every request to handle a situation that only occurs in 2 specific situations: Enabling and disabling maintenance mode. I understand however that this approach is taken because Drupal does not offer an API for reacting to this change it itself.
- This code clears the page cache but what about the Varnish cache. Varnish is supported so I would expect this to be cleared as well. However that logic belongs in the `ding_varnish` module. [The two modules cannot interact using hooks during `hook_boot`](https://www.drupal.org/node/496170). That leads to duplication of this logic in the two modules.

Due to these two issues I have looked into alternate solutions. One would be the [Maintenance Module API module](https://www.drupal.org/project/maintenance_mode_api). It provides a hook `hook_change_maintenance_mode` which is triggered when the maintenance module variable is set using the administration interface and through Drush.

`ding_frontend` and `ding_varnish` could implement these to clear their respective caches when changes occur.

**Comment by JrgenGNielsen** [comment]
@kasperg : I agree that the a hook_change_maintenance_mode is what's needed. I'll make an implementation of the Maintenance Module API ASAP.

**Comment by JrgenGNielsen** [comment]
@kasperg : Updated code. 
See also: [https://github.com/ding2/ding2/pull/23] & [https://github.com/ding2/ding_varnish/pull/1]

**Comment by kasperg** [comment]
Reviewed and approved. Thanks!

**Pull request merged**